changeset.go

```diff
@@ -109,63 +109,94 @@ func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetNa
 			}
 			return trace.Wrap(err)
 		}
-		err = cs.upsertResource(ctx, changesetNamespace, changesetName, raw.Raw)
+		err = cs.upsertResource(ctx, changesetNamespace, changesetName, id, raw.Raw)
 		if err != nil {
 			return trace.Wrap(err)
 		}
 	}
 }
 
-func (cs *Changeset) upsertResource(ctx context.Context, changesetNamespace, changesetName string, data []byte) error {
+func (cs *Changeset) upsertResource(ctx context.Context, changesetNamespace, changesetName, id string, data []byte) error {
 	tr, err := cs.createOrRead(changesetNamespace, changesetName, ChangesetSpec{Status: ChangesetStatusInProgress})
 	if err != nil {
 		return trace.Wrap(err)
 	}
 	if tr.Spec.Status != ChangesetStatusInProgress {
 		return trace.CompareFailed("cannot update changeset - expected status %q, got %q", ChangesetStatusInProgress, tr.Spec.Status)
 	}
-	var kind metav1.TypeMeta
-	err = yaml.NewYAMLOrJSONDecoder(bytes.NewReader(data), DefaultBufferSize).Decode(&kind)
+	headers, err := ParseResourceHeader(bytes.NewReader(data))
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	switch kind.Kind {
+
+	comp, err := cs.isOpCompleted(ctx, tr, id)
+	if comp {
+		log := log.WithFields(log.Fields{
+			"cs": tr.String(),
+		})
+		log.Infof("%s '%s' is already updated.", headers.TypeMeta.Kind, headers.ObjectMeta.Name)
+		return nil
+	}
+
+	if err != nil {
```

Comment:
Not sure I understand this logic. So if the resource hasn't been upserted yet, shouldn't we go ahead and upsert it below? Why are we calling the status here? And what to do if the status fails and returns that "operation not completed" error below?

Reply:
We can't upsert it again, because the operation was already started. If you start it for the second time you will lose original resource. And technically I have no solution yet how to handle error in status. Probably it should be the operation revert + upsert again. But it will not help in cases when a new resource can't start because of taints/node disk/cpu/memory pressures and etc. Looks like error should be handled manually.

Reply:
The scenario which I am trying to solve here:
1. I am running `rig upsert` for specific resource. It finished fine, resource is update.
2. Later my script failed somewhere and I need to restart the whole script - typical case for update hook, right?
3. I run it for the second time and if I do upsert again my original resource will be overwritten by the resource which I updated in step one. I lost the ability to revert everything back because I don't have original resources anymore.

My approach in this PR:
1. Update resource if `rig upsert ` for that specific resource(Kind, Name, Namespace) was invoked for the first time.
2. Skip the update for the all next invocation for the same resource.

That will allow me to keep original resources and I can rollback if needed.